### ISSUE TITLE
Still bind command even package is not installed

### ISSUE THREAD

**Issue by gongliyu** [jwiegley/use-package#437]
I'm not sure whether this is an issue. But I expect that no configuration should execute until the package is installed. 

For example,

(use-package swiper
  :bind (("C-s" . swiper)))

if I did not install swiper, then I can still use C-s to search. However, i got an error "Cannot open load file: No such file or directory, swiper"

My current workaroud is

(use-package swiper
  :if (package-installed-p 'swiper)
  :bind (("C-s" . swiper)))

Thanks!

**Comment by basil-conto** [comment]
``` emacs-lisp
(use-package swiper
  :bind (("C-s" . swiper)))
```

What this code effectively says is, "bind the key sequence `C-s` to a function called `swiper` which you can find in a file called `swiper.el` or `swiper.elc`". Specifically, this is done via a mechanism called [autoloading](https://www.gnu.org/software/emacs/manual/html_node/eintr/Autoload.html).

When the `swiper` function (or any of its associated key bindings) is invoked for the first time, the file that supposedly defines it (here listed as `swiper.el[c]`) is searched for in the directories listed in your `load-path` variable.

In other words, the `use-package` form above does not guarantee anything about the packages installed on your system and their whereabouts; it merely indicates what the required file will be named.

If you want `use-package` to automatically install the `swiper` package using the `package.el` package manager, you can use the `:ensure` keyword like so:

``` emacs-lisp
(use-package swiper
  :ensure
  :bind (("C-s" . swiper)))
```

Note that `swiper` is not available on [ELPA](https://elpa.gnu.org/packages/), so you will have to add [MELPA](https://melpa.org/packages/) to your `package-archives` variable.

If you do not want `use-package` to install the package automatically, but still wish for the binding to be guarded, your current workaround seems reasonable to me.

**Comment by gongliyu** [comment]
Thanks you!

**Comment by wyuenho** [comment]
@basil-conto What if I want the exact opposite behavior you've described, i.e. bind only after the package has successfully been loaded? My interpretation of the README is no code besides `:init` will be executed before the package has loaded. Experimentation shows this doesn't seem to be the case, is there a way to do it without putting `:if (package-installed-p 'whatever)` all over the place?

**Comment by basil-conto** [comment]
@wyuenho 
> What if I want the exact opposite behavior you've described, i.e. bind only after the package has successfully been loaded?

The `:config` clause registers code to be executed after successful loading of a feature:

```el
(use-package foo
  :defer
  :config
  (bind-key ...))
```

> My interpretation of the README is no code besides :init will be executed before the package has loaded

Depends on what you mean by "code". The `:init` clause is certainly the canonical approach, but `:preface` also executes arbitrary user code. Furthermore, the conditional clauses `:if`/`:when`/`:unless` also accept arbitrary forms, though their semantics are quite different.

From another perspective, the whole expansion of a `use-package` form could technically be executed before a feature is loaded.

> Experimentation shows this doesn't seem to be the case

Could you please elaborate and/or give an example?

> is there a way to do it without putting `:if (package-installed-p 'whatever)` all over the place

Does "it" refer to binding a key only after successful loading of a package? If not, could you please elaborate?

**Comment by wyuenho** [comment]
Yes I mean eval the form inside `:bind` only after the package has been loaded, it's not currently and I'm beginning to think it's a bug. Lots of people bind M-x to smex, if I'm moving my config to a new machine without that package installed, I'd like it to be unbound. `:ensure` doesn't work well with Melpa mirrors.

**Comment by basil-conto** [comment]
@wyuenho 

>  eval the form inside `:bind`

Which form? The `:bind` clause is intended for mapping keys to commands, not evaluating arbitrary forms. For something more flexible, consider using the `bind-key` macro.

> it's not currently and I'm beginning to think it's a bug

I don't think anyone can you help you without more information. If you have come across a reproducible issue, please consider opening a new issue where you describe the steps involved and the expected and actual outcomes.

> if I'm moving my config to a new machine without that package installed, I'd like it to be unbound

There are various ways you can guard the setting of a key binding, e.g.

```el
(when (some-condition)
  (bind-key ...)))
```

or, using `bind-key`'s optional predicate parameter,

```el
(bind-key <key-name> <command> <keymap> (some-condition))
```

> :ensure doesn't work well with Melpa mirrors

Again, this description is too vague and possibly unrelated to `use-package`. If you really think there is an issue with the expansion of the `:ensure` clause, and not `package.el`, `MELPA` or your configuration thereof, please consider opening a new issue with more information.

**Comment by wyuenho** [comment]
```elisp
(use-package smex
  :bind (("M-x" . smex))
  :config (smex-initialize))
```

If smex is not in load path, this form will still bind the key to the nonexistent command `smex`. There will be an error message in the minibuffer that says: "Cannot open load file: No such file or directory, smex".

Essentially, exactly the same situation as the description of this issue. I just reread the docs again, it seems that `:bind` is lifted to `(bind-keys)` inside `:init`. What I'm trying to understand is, why? Why is binding keys before a package has been loaded the default behavior?

**Comment by npostavs** [comment]
> Why is binding keys before a package has been loaded the default behavior?

The original point of `:bind` is to make an autoloaded command, so that the elisp is only loaded when you actually first use it (package.el creates these autoloads automatically, which makes the `:bind` half redundant, but `use-package` was not originally written with package.el in mind).

Somewhat related (it's more about `:init`, but a similar idea): #306, though the discussion was inconclusive, as I summarized here: https://github.com/jwiegley/use-package/issues/363#issuecomment-229738796.
